File: src/java/3_StructuralPatterns/4_Decorator/DecoratorExercise.java
package com.activemesa.structural.decorator.exercise;

class Bird
{
  public int age;

  public String fly()
  {
    return age < 10 ? "flying" : "too old";
  }
}

class Lizard
{
  public int age;

  public String crawl()
  {
    return (age > 1) ? "crawling" : "too young";
  }
}

class Dragon
{
  private int age;
  private Bird bird = new Bird();
  private Lizard lizard = new Lizard();
  public void setAge(int age)
  {
    this.age = bird.age = lizard.age = age;
  }
  public String fly()
  {
    return bird.fly();
  }
  public String crawl()
  {
    return lizard.crawl();
  }
}

File: src/js/3_StructuralPatterns/4_Decorator/decorator_exercise.js
class Bird
{
  constructor(age=0)
  {
    this.age = age;
  }

  fly()
  {
    return this.age < 10 ? 'flying' : 'too old';
  }
}

class Lizard
{
  constructor(age=0)
  {
    this.age = age;
  }

  crawl()
  {
    return this.age > 1 ? 'crawling' : 'too young';
  }
}

class Dragon
{
  constructor(age=0)
  {
    this._bird = new Bird();
    this._lizard = new Lizard();
    this._age = age;
  }

  set age(value) {
    this._age = this._bird.age
      = this._lizard.age = value;
  }

  get age() { return this._age; }

  fly() { return this._bird.fly(); }
  crawl() { return this._lizard.crawl(); }
}

describe('decorator', function()
{
  it('should let dragon fly or crawl', function()
  {
    let dragon = new Dragon();

    expect(dragon.fly()).toEqual('flying');
    expect(dragon.crawl()).toEqual('too young');

    dragon.age = 20;

    expect(dragon.fly()).toEqual('too old');
    expect(dragon.crawl()).toEqual('crawling');
  });
});